File: C#/0086. Partition List.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode Partition(ListNode head, int x) {
        ListNode smallNode=new ListNode(0);
        ListNode largerNode=new ListNode(0);
        ListNode pSmall=smallNode;
        ListNode pLarger=largerNode;
        while(head!=null){
            if(head.val>=x){
                pLarger.next=head;
                pLarger=pLarger.next;
                head=head.next;
            }
            else{
                pSmall.next=head;
                pSmall=pSmall.next;
                head=head.next;   
            }
        }
        pLarger.next=null;
        pSmall.next=largerNode.next;
        return smallNode.next;
    }
}

File: 20200215Python-China/0086. Partition List.py
'''
Given a linked list and a value x, partition it such that all nodes less than x come before nodes greater than or equal to x.

You should preserve the original relative order of the nodes in each of the two partitions.

Example:

Input: head = 1->4->3->2->5->2, x = 3
Output: 1->2->2->4->3->5

'''
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def partition(self, head: ListNode, x: int) -> ListNode:
        node1,node2=ListNode(0),ListNode(0)
        p_larger,p_small=node1,node2
        node1.next,node2.next=head,head
        while head:
            if head.val<x:
                p_small.next=head
                head=head.next
                p_small=p_small.next
            else:
                p_larger.next=head
                head=head.next
                p_larger=p_larger.next
        p_larger.next=None
        p_small.next=node1.next
        return node2.next
